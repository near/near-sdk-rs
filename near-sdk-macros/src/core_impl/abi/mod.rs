#[cfg(feature = "__abi-embed")]
mod abi_embed;
#[cfg(feature = "__abi-embed")]
pub use abi_embed::embed;

#[cfg(feature = "__abi-generate")]
mod abi_generator;
#[cfg(feature = "__abi-generate")]
pub use abi_generator::generate;
