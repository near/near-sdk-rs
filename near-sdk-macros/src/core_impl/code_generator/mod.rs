mod attr_sig_info;
pub use attr_sig_info::*;

mod impl_item_method_info;
pub use impl_item_method_info::*;

mod item_trait_info;
pub use item_trait_info::*;

mod trait_item_method_info;
pub use trait_item_method_info::*;

mod item_impl_info;
pub use item_impl_info::*;

mod item_struct_info;
pub use item_struct_info::*;
