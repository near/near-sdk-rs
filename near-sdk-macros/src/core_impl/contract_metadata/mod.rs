use darling::{ast::NestedMeta, Error, FromMeta};
use proc_macro2::TokenStream;
use quote::quote;

#[derive(FromMeta)]
struct MacroConfig {
    contract_metadata: Option<ContractMetadata>,
}

#[derive(serde::Serialize, Default, FromMeta)]
struct ContractMetadata {
    version: Option<String>,
    link: Option<String>,
    #[darling(multiple, rename = "standard")]
    standards: Vec<Standard>,
}

#[derive(FromMeta, serde::Serialize)]
struct Standard {
    standard: String,
    version: String,
}

impl ContractMetadata {
    fn populate(mut self) -> Self {
        if self.version.is_none() {
            let version = std::env::var("CARGO_PKG_VERSION").unwrap_or(String::from(""));
            if !version.is_empty() {
                self.version = Some(version);
            }
        }

        if self.link.is_none() {
            let repo = std::env::var("CARGO_PKG_REPOSITORY").unwrap_or(String::from(""));
            if !repo.is_empty() {
                self.link = Some(repo);
            }
        }

        // adding nep330 if it is not present
        if self.standards.is_empty()
            || self.standards.iter().all(|s| s.standard.to_ascii_lowercase() != "nep330")
        {
            self.standards
                .push(Standard { standard: "nep330".to_string(), version: "1.1.0".to_string() });
        }

        self
    }
}

/// Allows for the injection of the contract source metadata infomation into the contract code as
/// a constant.
pub(crate) fn contract_source_metadata_const(attr: proc_macro::TokenStream) -> TokenStream {
    if attr.to_string().is_empty() {
        let metadata = serde_json::to_string(&ContractMetadata::default().populate())
            .expect("ContractMetadata implements Serialize");

        return quote! {
           pub const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
        };
    }

    let attr_args = match NestedMeta::parse_meta_list(attr.into()) {
        Ok(v) => v,
        Err(e) => {
            return Error::from(e).write_errors();
        }
    };

    let args = match MacroConfig::from_list(&attr_args) {
        Ok(v) => v,
        Err(e) => {
            return e.write_errors();
        }
    };

    let metadata = serde_json::to_string(
        &args
            .contract_metadata
            .expect("Attribute input must be present given standard was followed")
            .populate(),
    )
    .expect("ContractMetadata implements Serialize");

    quote! {
        const CONTRACT_SOURCE_METADATA: &'static str = #metadata;
    }
}
