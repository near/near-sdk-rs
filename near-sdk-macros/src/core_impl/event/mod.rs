use proc_macro::TokenStream;
use proc_macro2::Span;
use quote::quote;
use syn::{parse_quote, ItemEnum, LitStr};

/// this function is used to inject serialization macros and the `near_sdk::EventMetadata` macro.
/// In addition, this function extracts the event's `standard` value and injects it as a constant to be used by
/// the `near_sdk::EventMetadata` derive macro
pub(crate) fn near_events(attr: TokenStream, item: TokenStream) -> TokenStream {
    // get standard from attr args
    let standard = get_standard_arg(&syn::parse_macro_input!(attr as syn::AttributeArgs));
    if standard.is_none() {
        return TokenStream::from(
            syn::Error::new(
                Span::call_site(),
                "Near events must have a `standard` value as an argument for `event_json` in the `near_bindgen` arguments. The value must be a string literal, e.g. \"nep999\", \"mintbase-marketplace\".",
            )
            .to_compile_error(),
        );
    }

    if let Ok(mut input) = syn::parse::<ItemEnum>(item) {
        let name = &input.ident;
        let standard_name = format!("{}_event_standard", name);
        let standard_ident = syn::Ident::new(&standard_name, Span::call_site());
        // NearEvent Macro handles implementation
        input
            .attrs
            .push(parse_quote! (#[derive(near_sdk::serde::Serialize, near_sdk::EventMetadata)]));
        input.attrs.push(parse_quote! (#[serde(crate="near_sdk::serde")]));
        input.attrs.push(parse_quote! (#[serde(tag = "event", content = "data")]));
        input.attrs.push(parse_quote! (#[serde(rename_all = "snake_case")]));

        TokenStream::from(quote! {
            const #standard_ident: &'static str = #standard;
            #input
        })
    } else {
        TokenStream::from(
            syn::Error::new(
                Span::call_site(),
                "`#[near_bindgen(event_json(standard = \"nepXXX\"))]` can only be used as an attribute on enums.",
            )
            .to_compile_error(),
        )
    }
}

/// This function returns the `version` value from `#[event_version("x.x.x")]`.
/// used by `near_sdk::EventMetadata`
pub(crate) fn get_event_version(var: &syn::Variant) -> Option<LitStr> {
    for attr in var.attrs.iter() {
        if attr.path.is_ident("event_version") {
            return attr.parse_args::<LitStr>().ok();
        }
    }
    None
}

/// this function returns the `standard` value from `#[near_bindgen(event_json(standard = "nepXXX"))]`
fn get_standard_arg(args: &[syn::NestedMeta]) -> Option<LitStr> {
    let mut standard: Option<LitStr> = None;
    for arg in args.iter() {
        if let syn::NestedMeta::Meta(syn::Meta::List(syn::MetaList { path, nested, .. })) = arg {
            if path.is_ident("event_json") {
                for event_arg in nested.iter() {
                    if let syn::NestedMeta::Meta(syn::Meta::NameValue(syn::MetaNameValue {
                        path,
                        lit: syn::Lit::Str(value),
                        ..
                    })) = event_arg
                    {
                        if path.is_ident("standard") {
                            standard = Some(value.to_owned());
                            break;
                        }
                    }
                }
            }
        }
    }
    standard
}
