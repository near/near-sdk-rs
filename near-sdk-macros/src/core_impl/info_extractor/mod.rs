use syn::{Receiver, ReturnType, Type};

mod serializer_attr;
pub use serializer_attr::SerializerAttr;

mod arg_info;
pub use arg_info::{ArgInfo, BindgenArgType};

mod handle_result_attr;
pub use handle_result_attr::HandleResultAttr;

mod attr_sig_info;
pub use attr_sig_info::AttrSigInfo;

mod impl_item_method_info;
pub use impl_item_method_info::ImplItemMethodInfo;

mod trait_item_method_info;
pub use trait_item_method_info::*;

mod item_trait_info;
pub use item_trait_info::ItemTraitInfo;

mod item_impl_info;

mod init_attr;
pub use init_attr::InitAttr;

mod visitor;

pub use item_impl_info::ItemImplInfo;

/// Type of serialization we use.
#[derive(Clone, PartialEq, Eq)]
#[allow(clippy::upper_case_acronyms)]
pub enum SerializerType {
    JSON,
    Borsh,
}

#[derive(Clone, PartialEq, Eq)]
pub enum MethodKind {
    Call(CallMethod),
    View(ViewMethod),
    Init(InitMethod),
}

#[derive(Clone, PartialEq, Eq)]
pub struct CallMethod {
    /// Whether method accepting $NEAR.
    pub is_payable: bool,
    /// Whether method can accept calls from self (current account)
    pub is_private: bool,
    /// The serializer that we use for the return type.
    pub result_serializer: SerializerType,
    /// The receiver, like `mut self`, `self`, `&mut self`, `&self`, or `None`.
    pub receiver: Option<Receiver>,
}

#[derive(Clone, PartialEq, Eq)]
pub struct ViewMethod {
    /// Whether method can accept calls from self (current account)
    pub is_private: bool,
    /// The serializer that we use for the return type.
    pub result_serializer: SerializerType,
    /// The receiver, like `mut self`, `self`, `&mut self`, `&self`, or `None`.
    pub receiver: Option<Receiver>,
}

#[derive(Clone, PartialEq, Eq)]
pub struct InitMethod {
    /// Whether method accepting $NEAR.
    pub is_payable: bool,
    /// Whether init method ignores state
    pub ignores_state: bool,
}

#[derive(Clone, PartialEq, Eq)]
pub struct Returns {
    /// The original return type of the method in the Rust AST.
    pub original: ReturnType,
    /// The return type of the method in our logic.
    pub kind: ReturnKind,
}

#[derive(Clone, PartialEq, Eq)]
pub enum ReturnKind {
    Default,
    General(Type),
    HandlesResult(Type),
}
