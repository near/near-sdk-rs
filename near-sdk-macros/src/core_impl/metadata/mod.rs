pub(crate) mod metadata_generator;
pub(crate) mod metadata_visitor;
