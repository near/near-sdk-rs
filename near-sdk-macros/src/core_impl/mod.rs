mod code_generator;
mod info_extractor;
mod metadata;
pub use code_generator::*;
pub use info_extractor::*;
pub use metadata::metadata_visitor::MetadataVisitor;
