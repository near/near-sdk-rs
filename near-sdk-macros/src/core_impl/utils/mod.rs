use proc_macro2::{Group, Span, TokenStream as TokenStream2, TokenTree};
use quote::quote;
use syn::spanned::Spanned;
use syn::token::{And, Mut};
use syn::{GenericArgument, Path, PathArguments, Signature, Type};

#[cfg(test)]
pub mod test_helpers;

/// Checks whether the given path is literally "Result".
/// Note that it won't match a fully qualified name `core::result::Result` or a type alias like
/// `type StringResult = Result<String, String>`.
pub(crate) fn path_is_result(path: &Path) -> bool {
    path.leading_colon.is_none()
        && path.segments.len() == 1
        && path.segments.iter().next().unwrap().ident == "Result"
}

/// Equivalent to `path_is_result` except that it works on `Type` values.
pub(crate) fn type_is_result(ty: &Type) -> bool {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() => path_is_result(&type_path.path),
        _ => false,
    }
}

/// Extracts the Ok type from a `Result` type.
///
/// For example, given `Result<String, u8>` type it will return `String` type.
pub(crate) fn extract_ok_type(ty: &Type) -> Option<&Type> {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() && path_is_result(&type_path.path) => {
            // Get the first segment of the path (there should be only one, in fact: "Result"):
            let type_params = &type_path.path.segments.first()?.arguments;
            // We are interested in the first angle-bracketed param responsible for Ok type ("<String, _>"):
            let generic_arg = match type_params {
                PathArguments::AngleBracketed(params) => Some(params.args.first()?),
                _ => None,
            }?;
            // This argument must be a type:
            match generic_arg {
                GenericArgument::Type(ty) => Some(ty),
                _ => None,
            }
        }
        _ => None,
    }
}

/// Checks whether the given path is literally "Vec".
/// Note that it won't match a fully qualified name `std::vec::Vec` or a type alias like
/// `type MyVec = Vec<String>`.
#[cfg(feature = "__abi-generate")]
fn path_is_vec(path: &Path) -> bool {
    path.leading_colon.is_none()
        && path.segments.len() == 1
        && path.segments.iter().next().unwrap().ident == "Vec"
}

/// Extracts the inner generic type from a `Vec<_>` type.
///
/// For example, given `Vec<String>` this function will return `String`.
#[cfg(feature = "__abi-generate")]
pub(crate) fn extract_vec_type(ty: &Type) -> Option<&Type> {
    match ty {
        Type::Path(type_path) if type_path.qself.is_none() && path_is_vec(&type_path.path) => {
            let type_params = &type_path.path.segments.first()?.arguments;
            let generic_arg = match type_params {
                // We are interested in the first (and only) angle-bracketed param:
                PathArguments::AngleBracketed(params) if params.args.len() == 1 => {
                    Some(params.args.first()?)
                }
                _ => None,
            }?;
            match generic_arg {
                GenericArgument::Type(ty) => Some(ty),
                _ => None,
            }
        }
        _ => None,
    }
}

/// Extracts reference and mutability tokens from a `Type` object. Also, strips top-level lifetime binding if present.
pub(crate) fn extract_ref_mut(ty: &Type) -> syn::Result<(Option<And>, Option<Mut>, Type)> {
    match ty {
        x @ (Type::Array(_) | Type::Path(_) | Type::Tuple(_) | Type::Group(_)) => {
            Ok((None, None, (*x).clone()))
        }
        Type::Reference(r) => Ok((Some(r.and_token), r.mutability, (*r.elem.as_ref()).clone())),
        _ => Err(syn::Error::new_spanned(ty, "Unsupported contract API type.")),
    }
}

/// Checks that the method signature is supported in the NEAR Contract API.
pub(crate) fn sig_is_supported(sig: &Signature) -> syn::Result<()> {
    if sig.asyncness.is_some() {
        return Err(syn::Error::new(sig.span(), "Contract API is not allowed to be async."));
    }
    if sig.abi.is_some() {
        return Err(syn::Error::new(
            sig.span(),
            "Contract API is not allowed to have binary interface.",
        ));
    }
    if sig.variadic.is_some() {
        return Err(syn::Error::new(
            sig.span(),
            "Contract API is not allowed to have variadic arguments.",
        ));
    }

    Ok(())
}

fn _sanitize_self(typ: TokenStream2, replace_with: &TokenStream2) -> (TokenStream2, Vec<Span>) {
    let mut self_occurrences = vec![];
    let trees = typ.into_iter().map(|t| match t {
        TokenTree::Ident(ident) if ident == "Self" => {
            self_occurrences.push(ident.span());
            replace_with
                .clone()
                .into_iter()
                .map(|mut t| {
                    t.set_span(ident.span());
                    t
                })
                .collect::<TokenStream2>()
        }
        TokenTree::Group(group) => {
            let (stream, self_occurrences_inner) = _sanitize_self(group.stream(), replace_with);
            self_occurrences.extend(self_occurrences_inner);
            TokenTree::Group(Group::new(group.delimiter(), stream)).into()
        }
        rest => rest.into(),
    });
    (trees.collect(), self_occurrences)
}

pub fn sanitize_self(typ: &Type, replace_with: &TokenStream2) -> syn::Result<SanitizeSelfResult> {
    let (ty_tokens, self_occurrences) = _sanitize_self(quote! { #typ }, replace_with);

    let ty = syn::parse2(ty_tokens).map_err(|original| {
        syn::Error::new(original.span(), "`Self` sanitization failed. Please report this as a bug.")
    })?;

    Ok(SanitizeSelfResult { ty, self_occurrences })
}

pub struct SanitizeSelfResult {
    pub ty: Type,
    pub self_occurrences: Vec<Span>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sanitize_self_works() {
        let typ: Type = syn::parse_str("Self").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "MyType");

        let typ: Type = syn::parse_str("Vec<Self>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "Vec < MyType >");

        let typ: Type = syn::parse_str("Vec<Vec<Self>>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(quote! { #sanitized }.to_string(), "Vec < Vec < MyType > >");

        let typ: Type = syn::parse_str("Option<[(Self, Result<Self, ()>); 2]>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        let sanitized = sanitize_self(&typ, &replace_with).unwrap().ty;
        assert_eq!(
            quote! { #sanitized }.to_string(),
            "Option < [(MyType , Result < MyType , () >) ; 2] >"
        );
    }

    #[test]
    fn sanitize_self_keeps_track_of_replacements() {
        let typ: Type = syn::parse_str("Self").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert_eq!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.len(), 1);

        let typ: Type = syn::parse_str("SomeType").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.is_empty());

        let typ: Type = syn::parse_str("Option<[(Self, Result<Self, ()>); 2]>").unwrap();
        let replace_with: TokenStream2 = syn::parse_str("MyType").unwrap();
        assert_eq!(sanitize_self(&typ, &replace_with).unwrap().self_occurrences.len(), 2);
    }
}
