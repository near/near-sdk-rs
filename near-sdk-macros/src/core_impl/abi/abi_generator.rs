use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::{format_ident, quote};
use syn::{parse_quote, Attribute, Expr, Lit::Str, Meta::NameValue, MetaNameValue, Type};

use crate::core_impl::{
    utils, BindgenArgType, ImplItemMethodInfo, ItemImplInfo, MethodKind, ReturnKind, SerializerType,
};

pub fn generate(i: &ItemImplInfo) -> TokenStream2 {
    if i.methods.is_empty() {
        // Short-circuit if there are no public functions to export to ABI
        return TokenStream2::new();
    }

    let functions: Vec<TokenStream2> = i.methods.iter().map(|m| m.abi_struct()).collect();
    let first_function_name = &i.methods[0].attr_signature_info.ident;
    let near_abi_symbol = format_ident!("__near_abi_{}", first_function_name);
    quote! {
        #[cfg(not(target_arch = "wasm32"))]
        const _: () = {
            #[no_mangle]
            pub extern "C" fn #near_abi_symbol() -> (*const u8, usize) {
                use ::std::string::String;

                let mut gen = ::near_sdk::schemars::gen::SchemaGenerator::default();
                let functions = vec![#(#functions),*];
                let mut data = ::std::mem::ManuallyDrop::new(
                    ::near_sdk::serde_json::to_vec(&::near_sdk::__private::ChunkedAbiEntry::new(
                        functions,
                        gen.into_root_schema_for::<String>(),
                    ))
                    .unwrap(),
                );
                data.shrink_to_fit();
                assert!(data.len() == data.capacity());
                (data.as_ptr(), data.len())
            }
        };
    }
}

impl ImplItemMethodInfo {
    /// Generates ABI struct for this function.
    ///
    /// # Example:
    /// The following function:
    /// ```ignore
    /// /// I am a function.
    /// #[handle_result]
    /// pub fn f3(&mut self, arg0: FancyStruct, arg1: u64) -> Result<IsOk, Error> { }
    /// ```
    /// will produce this struct:
    /// ```ignore
    /// near_sdk::__private::AbiFunction {
    ///     name: "f3".to_string(),
    ///     doc: Some(" I am a function.".to_string()),
    ///     kind: near_sdk::__private::AbiFunctionKind::Call,
    ///     modifiers: vec![],
    ///     params: near_sdk::__private::AbiParameters::Json {
    ///         args: vec![
    ///             near_sdk::__private::AbiJsonParameter {
    ///                 name: "arg0".to_string(),
    ///                 type_schema: gen.subschema_for::<FancyStruct>(),
    ///             },
    ///             near_sdk::__private::AbiJsonParameter {
    ///                 name: "arg1".to_string(),
    ///                 type_schema: gen.subschema_for::<u64>(),
    ///             }
    ///         ]
    ///     },
    ///     callbacks: vec![],
    ///     callbacks_vec: None,
    ///     result: Some(near_sdk::__private::AbiType::Json {
    ///         type_schema: gen.subschema_for::<IsOk>(),
    ///     })
    /// }
    /// ```
    /// If args are serialized with Borsh it will not include `#[derive(::near_sdk::borsh::BorshSchema)]`.
    pub fn abi_struct(&self) -> TokenStream2 {
        let attr_signature_info = &self.attr_signature_info;

        let function_name_str = attr_signature_info.ident.to_string();
        let function_doc = match parse_rustdoc(&attr_signature_info.non_bindgen_attrs) {
            Some(doc) => quote! { ::std::option::Option::Some(::std::string::String::from(#doc)) },
            None => quote! { ::std::option::Option::None },
        };
        let mut modifiers = vec![];
        let kind = match &attr_signature_info.method_kind {
            MethodKind::View(_) => quote! { ::near_sdk::__private::AbiFunctionKind::View },
            MethodKind::Call(_) => {
                quote! { ::near_sdk::__private::AbiFunctionKind::Call }
            }
            MethodKind::Init(_) => {
                modifiers.push(quote! { ::near_sdk::__private::AbiFunctionModifier::Init });
                quote! { ::near_sdk::__private::AbiFunctionKind::Call }
            }
        };
        if attr_signature_info.is_payable() {
            modifiers.push(quote! { ::near_sdk::__private::AbiFunctionModifier::Payable });
        }
        if attr_signature_info.is_private() {
            modifiers.push(quote! { ::near_sdk::__private::AbiFunctionModifier::Private });
        }
        let modifiers = quote! {
            ::std::vec![#(#modifiers),*]
        };

        let mut params = Vec::<TokenStream2>::new();
        let mut callbacks = Vec::<TokenStream2>::new();
        let mut callback_vec: Option<TokenStream2> = None;
        for arg in &attr_signature_info.args {
            let typ = &arg.ty;
            let arg_name = arg.ident.to_string();
            match arg.bindgen_ty {
                BindgenArgType::Regular => {
                    let schema = generate_schema(typ, &arg.serializer_ty);
                    match arg.serializer_ty {
                        SerializerType::JSON => params.push(quote! {
                            ::near_sdk::__private::AbiJsonParameter {
                                name: ::std::string::String::from(#arg_name),
                                type_schema: #schema,
                            }
                        }),
                        SerializerType::Borsh => params.push(quote! {
                            ::near_sdk::__private::AbiBorshParameter {
                                name: ::std::string::String::from(#arg_name),
                                type_schema: #schema,
                            }
                        }),
                    };
                }
                BindgenArgType::CallbackArg => {
                    callbacks.push(generate_abi_type(typ, &arg.serializer_ty));
                }
                BindgenArgType::CallbackResultArg => {
                    let typ = if let Some(ok_type) = utils::extract_ok_type(typ) {
                        ok_type
                    } else {
                        return syn::Error::new_spanned(
                            &arg.ty,
                            "Function parameters marked with \
                                #[callback_result] should have type Result<T, PromiseError>",
                        )
                        .into_compile_error();
                    };
                    callbacks.push(generate_abi_type(typ, &arg.serializer_ty));
                }
                BindgenArgType::CallbackArgVec => {
                    if callback_vec.is_none() {
                        let typ = if let Some(vec_type) = utils::extract_vec_type(typ) {
                            vec_type
                        } else {
                            return syn::Error::new_spanned(
                                &arg.ty,
                                "Function parameters marked with #[callback_vec] should have type Vec<T>",
                            )
                            .into_compile_error();
                        };

                        callback_vec = Some(self.abi_callback_vec_tokens(typ));
                    } else {
                        return syn::Error::new(
                            Span::call_site(),
                            "A function can only have one #[callback_vec] parameter.",
                        )
                        .to_compile_error();
                    }
                }
            };
        }
        let params = match attr_signature_info.input_serializer {
            SerializerType::JSON => quote! {
                ::near_sdk::__private::AbiParameters::Json {
                    args: ::std::vec![#(#params),*]
                }
            },
            SerializerType::Borsh => quote! {
                ::near_sdk::__private::AbiParameters::Borsh {
                    args: ::std::vec![#(#params),*]
                }
            },
        };
        let callback_vec = callback_vec.unwrap_or(quote! { ::std::option::Option::None });

        let result = self.abi_result_tokens();

        quote! {
             ::near_sdk::__private::AbiFunction {
                 name: ::std::string::String::from(#function_name_str),
                 doc: #function_doc,
                 kind: #kind,
                 modifiers: #modifiers,
                 params: #params,
                 callbacks: ::std::vec![#(#callbacks),*],
                 callbacks_vec: #callback_vec,
                 result: #result
             }
        }
    }

    fn abi_result_tokens(&self) -> TokenStream2 {
        use ReturnKind::*;

        match &self.attr_signature_info.returns.kind {
            Default => quote! { ::std::option::Option::None },
            General(ty) => self.abi_result_tokens_with_return_value(ty),
            HandlesResult(ty) => {
                // extract the `Ok` type from the result
                let ty = parse_quote! { <#ty as near_sdk::__private::ResultTypeExt>::Okay };
                self.abi_result_tokens_with_return_value(&ty)
            }
        }
    }

    fn abi_result_tokens_with_return_value(&self, return_value_type: &Type) -> TokenStream2 {
        use MethodKind::*;

        let some_abi_type = |result_serializer: &SerializerType| {
            let abi_type = generate_abi_type(return_value_type, result_serializer);
            quote! { ::std::option::Option::Some(#abi_type) }
        };

        match &self.attr_signature_info.method_kind {
            Call(call_method) => some_abi_type(&call_method.result_serializer),
            // Init methods don't return a value, they just save the newly created contract state.
            Init(_) => quote! { ::std::option::Option::None },
            View(view_method) => some_abi_type(&view_method.result_serializer),
        }
    }

    fn abi_callback_vec_tokens(&self, callback_vec_type: &Type) -> TokenStream2 {
        let abi_type = |result_serializer: &SerializerType| {
            let tokens = generate_abi_type(callback_vec_type, result_serializer);
            quote! {
                ::std::option::Option::Some(#tokens)
            }
        };

        match &self.attr_signature_info.method_kind {
            MethodKind::Call(call_method) => abi_type(&call_method.result_serializer),
            MethodKind::Init(_) => quote! { ::std::option::Option::None },
            MethodKind::View(view_method) => abi_type(&view_method.result_serializer),
        }
    }
}

fn generate_schema(ty: &Type, serializer_type: &SerializerType) -> TokenStream2 {
    match serializer_type {
        SerializerType::JSON => quote! {
            gen.subschema_for::<#ty>()
        },
        SerializerType::Borsh => quote! {
            ::near_sdk::borsh::schema_container_of::<#ty>()
        },
    }
}

fn generate_abi_type(ty: &Type, serializer_type: &SerializerType) -> TokenStream2 {
    let schema = generate_schema(ty, serializer_type);
    match serializer_type {
        SerializerType::JSON => quote! {
            ::near_sdk::__private::AbiType::Json {
                type_schema: #schema,
            }
        },
        SerializerType::Borsh => quote! {
            ::near_sdk::__private::AbiType::Borsh {
                type_schema: #schema,
            }
        },
    }
}

pub fn parse_rustdoc(attrs: &[Attribute]) -> Option<String> {
    let doc = attrs
        .iter()
        .filter_map(|attr| {
            if attr.path().is_ident("doc") {
                if let NameValue(MetaNameValue { value: Expr::Lit(value), .. }) = attr.meta.clone()
                {
                    if let Str(doc) = value.lit {
                        return Some(doc.value());
                    }
                }
            }
            None
        })
        .collect::<Vec<_>>()
        .join("\n");

    if doc.is_empty() {
        None
    } else {
        Some(doc)
    }
}

// Rustfmt removes comas.
#[rustfmt::skip]
#[cfg(test)]
mod tests {
    use proc_macro2::TokenStream;
    use syn::{parse_quote, Type};
    use crate::core_impl::ImplItemMethodInfo;
    use crate::core_impl::utils::test_helpers::{local_insta_assert_snapshot, pretty_print_syn_str};
    use quote::quote;


    fn pretty_print_fn_body_syn_str(input: TokenStream) -> String {
        let input =  quote!(
            fn main() {
            #input
            }
        );
        let res = pretty_print_syn_str(&input).unwrap();
       res.strip_prefix("fn main() {\n").unwrap().strip_suffix("}\n").unwrap().to_string()
    }
    
    #[test]
    fn test_generate_abi_fallible_json() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            /// I am a function.
            #[handle_result]
            pub fn f3(&mut self, arg0: FancyStruct, arg1: u64) -> Result<IsOk, Error> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();

        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }

    #[test]
    fn test_generate_abi_fallible_borsh() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            #[result_serializer(borsh)]
            #[payable]
            #[handle_result]
            pub fn f3(&mut self, #[serializer(borsh)] arg0: FancyStruct) -> Result<IsOk, Error> { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();

        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }
    
    #[test]
    fn test_generate_abi_private_callback_vec() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            #[private] 
            pub fn method(
                &self, 
                #[callback_vec] x: Vec<String>, 
            ) -> bool { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();
       
        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }
    
    #[test]
    fn test_generate_abi_callback_args() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            pub fn method(&self, #[callback_unwrap] #[serializer(borsh)] x: &mut u64, #[serializer(borsh)] y: String, #[callback_unwrap] #[serializer(json)] z: Vec<u8>) { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();

        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }
    
    #[test]
    fn test_generate_abi_init_ignore_state() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            #[init(ignore_state)]
            pub fn new() -> u64 { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();

        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }
    
    #[test]
    fn test_generate_abi_no_return() {
        let impl_type: Type = syn::parse_str("Test").unwrap();
        let mut method = parse_quote! {
            pub fn method() { }
        };
        let method_info = ImplItemMethodInfo::new(&mut method, false, impl_type).unwrap().unwrap();
        let actual = method_info.abi_struct();

        local_insta_assert_snapshot!(pretty_print_fn_body_syn_str(actual));
    }
}
