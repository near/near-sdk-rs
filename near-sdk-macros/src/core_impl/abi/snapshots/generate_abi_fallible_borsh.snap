---
source: near-sdk-macros/src/core_impl/abi/abi_generator.rs
expression: pretty_print_fn_body_syn_str(actual)
---
    ::near_sdk::__private::AbiFunction {
        name: ::std::string::String::from("f3"),
        doc: ::std::option::Option::None,
        kind: ::near_sdk::__private::AbiFunctionKind::Call,
        modifiers: ::std::vec![::near_sdk::__private::AbiFunctionModifier::Payable],
        params: ::near_sdk::__private::AbiParameters::Borsh {
            args: ::std::vec![
                ::near_sdk::__private::AbiBorshParameter { name :
                ::std::string::String::from("arg0"), type_schema :
                ::near_sdk::borsh::schema_container_of:: < FancyStruct > (), }
            ],
        },
        callbacks: ::std::vec![],
        callbacks_vec: ::std::option::Option::None,
        result: ::std::option::Option::Some(::near_sdk::__private::AbiType::Borsh {
            type_schema: ::near_sdk::borsh::schema_container_of::<
                <Result<IsOk, Error> as near_sdk::__private::ResultTypeExt>::Okay,
            >(),
        }),
    }

